### PR TITLE
Add config to set precision of torch fp32 matrix multiply

### PR DESCRIPTION
**Pull request by lgeiger**
This adds a config option to set the [precision of torch float32 matrix multiplications](https://docs.pytorch.org/docs/2.13/generated/torch.set_float32_matmul_precision.html) for eval and training.

The default value doesn't change behaviour. However, I evaluated the current checkpoint with `float32_matmul_precision="high"` and the difference is pretty much just noise.

On an A100 GPU, which has support for the TensorFloat32 (aka `float32_matmul_precision="high"`), end2end eval runtime decreases by only 8% since matrix multiplies are currently not the biggest bottleneck. However, transformer training throughput increases by 2.2x which is significant 🚀
Automatic bfloat16 mixed precision would improve training speed by another 10-15% but I haven't verified that this won't impact on the final accuracy after longer training.

### CODE REVIEW

File: retrochimera/cli/eval.py [hunk 1]
```diff
@@ -47,11 +49,12 @@ class BackwardModelConfig(ModelConfig):
 class EvalConfig(BackwardModelConfig, eval_syntheseus.BaseEvalConfig):
     """Config for running evaluation on a given dataset."""
 
-    pass
+    float32_matmul_precision: Float32MatmulPrecision = Float32MatmulPrecision.highest
```
Comment (lgeiger) [inline retrochimera/cli/eval.py:52]: Unfortunately, omegaconf doesn't support `typing.Literal` yet, so using an enum until they release their next version.

**Review by kmaziarz — APPROVED** [review]
LGTM.
